README.md
# Xanadu

Xanadu is a framework for developing Database Driven Web Apps. 

Xanadu uses PHP, MySQL, HTML, Bootstrap, CSS, and Javascript with several amazing includes.

We've been developing Database Apps for decades with FileMaker and Xojo. FileMaker rocks but licensing became too expensive. Xojo has fantastic pricing, but is perpetually buggy. We had to find an alternative platform. PHP kept coming up in our searches. Long story short, we decided PHP is the best option just to get back to an affordable and solid platform.

Xanadu is FREE because we want to help others. If there's enough interest, we hope to build a community of developers.

Read more about Xanadu: https://campsoftware.com/products/xanadu.php

ChangeLog: https://github.com/campsoftware/xanadu/blob/master/CHANGELOG.md

CHANGELOG.md
# Xanadu Change Log

2021-06-12-15-47-52
- Added \xan\microsecsTracker class for millisecond level time tracking of code and loops.

2021-06-10-16-22-27
- Added a Constant TWOFACTORAUTH_ENABLED to enable or disable 2FA.
- Added eleDBMetaRender ELE_AS_STRING.
- Added User table columns for '2FA via Phone' and '2FA via Email' to allow options for receiving the code. 
- Updated the Password Meter to check for Password Length and spread out the Levels ( Weak, Moderate, Strong ).
- Register and Change Passwords now require a Strong password based on the Password Meter.

2021-06-08-16-21-56
- Removed setting $_SESSION[ SESS_USER ] from init.php Session was not started yet. It's set in mmUsersT->doLogin.
- Moved setting $_SESSION[ SESS_URL ] to aloe/framework/session.php.
- Added a mb-1 to the button classes, ELE_CLASS_BUTTON_RG_PRIMARY, ELE_CLASS_BUTTON_SM_PRIMARY, ELE_CLASS_BUTTON_RG_SECONDARY, and ELE_CLASS_BUTTON_SM_SECONDARY to add bit of vertical space.

2021-06-08-15-26-13
- Fixed eleTable->render to correctly set table thead and tbody.
- Renamed FontAwesome properties to FontIcon since we're using FontAwesome and Boostrap Icons.
- $_SESSION[ 'urlCurrent' ] is now $_SESSION[ SESS_URL ] to replace the string with a constant.

2021-06-06-16-20-23
- Updated Init, mmSettingsT, and Settings Card values to appear in similar order.
- Added Settings for Phone Number, Email Address, and Postal Address.

2021-06-05-16-30-10
- Added colMeta->eleAlign that accepts left, right, center, justify which defaults to left.
- Added Utility Function arrayContainsString for searching arrays for a value.
- Added additional reqID properties 2, 3, 4, 5
- Added getListRow params $rowHaButton and $rowColNamesA for tables in List Cards.
- Added eleCard IDs to make Showing and Hiding Cards simple.
- Added eleSearchBarListDB->render optional properties for $inclColA, $inclColMod, and $inclColKeys.

2021-05-15-14-37-21
- Removed an unneeded call to fmREST.

2021-05-13-17-45-29
- Added Contacts Comms create SMS to go with call.

2021-05-13-17-12-47
- Fixed \xan\response->metaSet(). Removed an extra quote and closing bracket.

2021-05-13-17-01-49
- Updated Home to add more examples. Added a use of php date_default_timezone_get.

2021-05-13-16-39-14
- Added a clickable link to the 2FA SMS and Email messages that will open a new tab to complete the Login. Works similar to the Registration One Time Code.

2021-05-13-15-34-17
- Added setting the default timezone to UTC in init.php.
- Added a function for arrayValuesWrapWith and arrayValuesWrapWithBackticks. Using to wrapping Column Names with backticks. Moved array functions higher within functions_utility.php.
- Updated dbSQL_InsertOrUpdate to use arrayValuesWrapWithBackticks. Goal is to wrap all SQL Statement Column Names with backticks everywhere possible.

2021-05-13-14-38-07
- Renamed /sql/ files from sequential numbers to dates.
- Added /sql/xan-2021-05-13.sql to add columns for constants: APP_COUNTRY_CODE, TWITTER_SITE, TWITTER_AUTHOR.
- Changed ALTER TABLE commands to use AFTER to keep table column order.
- Changed init.php to use the database values for APP_COUNTRY_CODE, TWITTER_SITE, TWITTER_AUTHOR.

2021-05-13-12-57-31
- Added constants: APP_COUNTRY_CODE, APP_LOCALE, TWITTER_SITE, TWITTER_AUTHOR. Need to add to the Settings Database Record.
- Added properties to \xan\response to support optionally adding html meta tags via \xan\respone->metaSet();
- Added \xan\respone->metaSet(); to several pages.

2021-05-11-15-09-05
- Updated Register Password Rating as a function in xan.js xanPasswordRating.
- Added User Change Password Rating using xan.js xanPasswordRating.

2021-05-11-14-10-04
- Added a simple Password Rating of Weak in red, Moderate in yellow, Strong in green.

2021-05-11-12-51-44
- Added a function dbSQL_InsertOrUpdate that creates a sql statement that will INSERT / ON DUPLICATE KEY UPDATE that will either Insert a record or Update a record if it already exists.
- Renamed a few functions to: dbSQL_InsertValuesQuestions, dbSearchTerm_SQL, dbSearchTerm_BindNamesA, dbSearchTerm_BindValuesA, dbQueryBuilder_DataType, dbQueryBuilder_FindFilter, and dbQueryOrderBy_DropdownItem.

2021-05-10-12-18-05
- Fixed an issue with quotes when displaying font icons during ajax calls.

2021-05-06-17-52-02
- Added Bootstrap Icons.
- Replaced iconFA function with fontIcon. Adds a tags parameter. Works with both Bootstrap Icons and FontAwesome.
- Planning to remove FontAwesome when to Bootstrap Icons collection is larger.
- Replaced all FontAwesome html with the equivalent fontIcon function. 
- Ranamed all "FA_" icon constant names to "FI_".
- Removed eleCard->renderListItemLink as it's now redundant.

2021-05-04-16-46-46
- Commented out XanDo javascript alert.
- Touched images and index.php.

2021-05-04-15-48-16
- Summary: Simplifying List Cards while adding options: Items Text, Items Image + Text, or a Table Row.
- Removed three functions from moduleMeta class and each subclass. Replaced with sharable moduleMeta functions.
- Added colMeta property widthForTable that defaults to empty string but can be overridden per Table Column Name.
- Added two new moduleMeta getList functions: getListItem is for text, getListItemWImage is for an image + text, and getListRow is for html tables at full page width with a default column width.
- Updated content-cards.php Lists for Contacts, Contact Picker ( image + text ), Settings-Users ( text ), and APIRequests ( html table ). It's now easy to change how Lists are displayed.

2021-05-01-18-08-00
- Added a link to the ChangeLog in the ReadMe.

2021-05-01-17-58-33
- Updated APIRequests getListItemRowHeader and getListItemRow functions to a single function getListRow as an option to getListItem.
- Updated APIRequests List Card Styles from mini/wide to items/rows.

xan/xan.js
///////////////////////////////////////////////////////////
// Utility
function xanEventReturnFalse( e ) {
    return false;
}

///////////////////////////////////////////////////////////
// Messages

function xanMessageDisplay( selector, html, id, doFadeOutID, doFadeOut, doConsoleLog ) {
    if ( document.getElementById( "xanMessage" ) ) {
        let cell = document.getElementById( "xanMessage" ).insertCell( -1 );
        cell.innerHTML = "<span id='" + id + "'>" + html + "</span>";
        if ( doFadeOutID !== '' ) {
            $( "#" + doFadeOutID ).delay( 3000 ).fadeOut( 'slow' );
        }
        if ( doFadeOut === true ) {
            $( "#" + id ).delay( 3000 ).fadeOut( 'slow' );
        }
    }
    if ( doConsoleLog === true ) {
        console.log( xanConsoleCleanHTML( html ) );
    }
}

function xanConsoleDisplay() {
    for ( i = 0; i < xanConsoleMsgs.length; i++ ) {
        console.log( xanConsoleCleanHTML( xanConsoleMsgs[ i ] ) );
    }
    xanConsoleMsgs = [];
}

function xanConsoleCleanHTML( html ) {
    html = html.replaceAll( "<br />", " " );
    html = strip_tags( html );
    html = addslashes( html );
    html = html.replaceAll( "&nbsp;", " " );
    html = html.replace( /\s+/g, " " );
    html = html.trim();
    return html;
}

///////////////////////////////////////////////////////////
// Auto Logout
function xanSchedAutoLogout( schedID, secs ) {
    clearTimeout( schedID );
    schedID = setTimeout( function () {
        location.href = "/logout";
    }, secs * 1000 );
    console.log( "AutoLogout at " + moment().add( secs, 'seconds' ).format( 'hh:mm:ss A' ) );
    return schedID;
}

///////////////////////////////////////////////////////////
// PHP Equivalents

function addslashes( str ) {
    str = str.replace( /\\/g, '\\\\' );
    str = str.replace( /\'/g, '\\\'' );
    str = str.replace( /\"/g, '\\"' );
    str = str.replace( /\0/g, '\\0' );
    return str;
}

function stripslashes( str ) {
    str = str.replace( /\\'/g, '\'' );
    str = str.replace( /\\"/g, '"' );
    str = str.replace( /\\0/g, '\0' );
    str = str.replace( /\\\\/g, '\\' );
    return str;
}

function strip_tags( str ) {
    if ( ( str === null ) || ( str === '' ) )
        return false;
    else
        str = str.toString();
    return str.replace( /(<([^>]+)>)/ig, '' );
}

///////////////////////////////////////////////////////////
// Elements

function xanEleSelectOnChange( pIDSelect, pValue ) {
    hiddenSelector = $( "#" + pIDSelect );
    selectSelector = $( "#" + pIDSelect + "_Select" );
    modalSelector = $( "#" + pIDSelect + "_Modal" );
    switch ( pValue ) {
        case 'Other':
            selectSelector.val( selectSelector.data( 'prev' ) );
            modalSelector.modal( "show" );
            break;
        case 'Clear':
            selectSelector.val( "" );
            hiddenSelector.val( "" ).trigger( "change" );
            break;
        default:
            hiddenSelector.val( selectSelector.val() ).trigger( "change" );
    }
}

function xanEleSelectOtherAction( pIDSelect ) {
    hiddenSelector = $( "#" + pIDSelect );
    selectSelector = $( "#" + pIDSelect + "_Select" );
    modalInputSelector = $( "#" + pIDSelect + "_Modal_Input" );
    let theValue = modalInputSelector.val();
    let theSelect = document.getElementById( pIDSelect + "_Select" );
    let theOpt = document.createElement( "option" );
    theOpt.value = theValue;
    theOpt.innerHTML = theValue;
    theSelect.add( theOpt, theSelect[ 0 ] );
    selectSelector.val( theValue );
    hiddenSelector.val( theValue ).trigger( "change" );
}

function xanEleFlatpickrSet( pIDSelect, pIDFlatpickr, pTypeFlatpickr, pValue ) {
    // Select Reset
    document.querySelector( pIDSelect ).value = "";
    document.querySelector( pIDSelect ).blur();

    // Return if Empty
    if ( pValue === "" ) return;

    // Get Formats
    let formatMoment = "";
    let formatFlatpickr = "";
    switch ( pTypeFlatpickr ) {
        case "date":
            formatMoment = 'YYYY-MM-DD';
            formatFlatpickr = 'Y-m-d';
            break;
        case "datetime":
            formatMoment = 'YYYY-MM-DD HH:mm:ss';
            formatFlatpickr = 'Y-m-d H:i:S';
            break;
        case "time":
            formatMoment = 'HH:mm:ss';
            formatFlatpickr = 'H:i:S';
            break;
    }

    // Get Moment
    let eleMoment = '';
    switch ( pValue ) {
        case "Today":
            eleMoment = moment().format( formatMoment );
            break;
        case "Yesterday":
            eleMoment = moment().subtract( 1, "days" ).format( formatMoment );
            break;
        case "Tomorrow":
            eleMoment = moment().add( 1, "days" ).format( formatMoment );
            break;

        case "Year Begin":
            eleMoment = moment().startOf( 'year' ).format( formatMoment );
            break;
        case "Year End":
            eleMoment = moment().endOf( 'year' ).format( formatMoment );
            break;

        case "Q1 Begin":
            eleMoment = moment().quarter( 1 ).startOf( 'quarter' ).format( formatMoment );
            break;
        case "Q1 End":
            eleMoment = moment().quarter( 1 ).endOf( 'quarter' ).format( formatMoment );
            break;
        case "Q2 Begin":
            eleMoment = moment().quarter( 2 ).startOf( 'quarter' ).format( formatMoment );
            break;
        case "Q2 End":
            eleMoment = moment().quarter( 2 ).endOf( 'quarter' ).format( formatMoment );
            break;
        case "Q3 Begin":
            eleMoment = moment().quarter( 3 ).startOf( 'quarter' ).format( formatMoment );
            break;
        case "Q3 End":
            eleMoment = moment().quarter( 3 ).endOf( 'quarter' ).format( formatMoment );
            break;
        case "Q4 Begin":
            eleMoment = moment().quarter( 4 ).startOf( 'quarter' ).format( formatMoment );
            break;
        case "Q4 End":
            eleMoment = moment().quarter( 4 ).endOf( 'quarter' ).format( formatMoment );
            break;

        case "Month Begin":
            eleMoment = moment().startOf( 'month' ).format( formatMoment );
            break;
        case "Month End":
            eleMoment = moment().endOf( 'month' ).format( formatMoment );
            break;
        case "Last Month Begin":
            eleMoment = moment().subtract( 1, 'month' ).startOf( 'month' ).format( formatMoment );
            break;
        case "Last Month End":
            eleMoment = moment().subtract( 1, 'month' ).endOf( 'month' ).format( formatMoment );
            break;
        case "Next Month Begin":
            eleMoment = moment().add( 1, 'month' ).startOf( 'month' ).format( formatMoment );
            break;
        case "Next Month End":
            eleMoment = moment().add( 1, 'month' ).endOf( 'month' ).format( formatMoment );
            break;

        case "Week Begin":
            eleMoment = moment().startOf( 'week' ).format( formatMoment );
            break;
        case "Week End":
            eleMoment = moment().endOf( 'week' ).format( formatMoment );
            break;
        case "Last Week Begin":
            eleMoment = moment().subtract( 1, 'week' ).startOf( 'week' ).format( formatMoment );
            break;
        case "Last Week End":
            eleMoment = moment().subtract( 1, 'week' ).endOf( 'week' ).format( formatMoment );
            break;
        case "Next Week Begin":
            eleMoment = moment().add( 1, 'week' ).startOf( 'week' ).format( formatMoment );
            break;
        case "Next Week End":
            eleMoment = moment().add( 1, 'week' ).endOf( 'week' ).format( formatMoment );
            break;

        case "Now":
            eleMoment = moment().format( formatMoment );
            break;

        case "7am":
            eleMoment = moment( '07:00', 'hh:mm' ).format( formatMoment );
            break;
        case "8am":
            eleMoment = moment( '08:00', 'hh:mm' ).format( formatMoment );
            break;
        case "9am":
            eleMoment = moment( '09:00', 'hh:mm' ).format( formatMoment );
            break;
        case "10am":
            eleMoment = moment( '10:00', 'hh:mm' ).format( formatMoment );
            break;
        case "11am":
            eleMoment = moment( '11:00', 'hh:mm' ).format( formatMoment );
            break;
        case "12pm":
            eleMoment = moment( '12:00', 'hh:mm' ).format( formatMoment );
            break;

        case "1pm":
            eleMoment = moment( '13:00', 'hh:mm' ).format( formatMoment );
            break;
        case "2pm":
            eleMoment = moment( '14:00', 'hh:mm' ).format( formatMoment );
            break;
        case "3pm":
            eleMoment = moment( '15:00', 'hh:mm' ).format( formatMoment );
            break;
        case "4pm":
            eleMoment = moment( '16:00', 'hh:mm' ).format( formatMoment );
            break;
        case "5pm":
            eleMoment = moment( '17:00', 'hh:mm' ).format( formatMoment );
            break;
        case "6pm":
            eleMoment = moment( '18:00', 'hh:mm' ).format( formatMoment );
            break;
        case "7pm":
            eleMoment = moment( '19:00', 'hh:mm' ).format( formatMoment );
            break;

        case "8pm":
            eleMoment = moment( '20:00', 'hh:mm' ).format( formatMoment );
            break;
        case "9pm":
            eleMoment = moment( '21:00', 'hh:mm' ).format( formatMoment );
            break;
        case "10pm":
            eleMoment = moment( '22:00', 'hh:mm' ).format( formatMoment );
            break;
        case "11pm":
            eleMoment = moment( '23:00', 'hh:mm' ).format( formatMoment );
            break;
        case "12am":
            eleMoment = moment( '00:00', 'hh:mm' ).format( formatMoment );
            break;

        case "1am":
            eleMoment = moment( '01:00', 'hh:mm' ).format( formatMoment );
            break;
        case "2am":
            eleMoment = moment( '02:00', 'hh:mm' ).format( formatMoment );
            break;
        case "3am":
            eleMoment = moment( '03:00', 'hh:mm' ).format( formatMoment );
            break;
        case "4am":
            eleMoment = moment( '04:00', 'hh:mm' ).format( formatMoment );
            break;
        case "5am":
            eleMoment = moment( '05:00', 'hh:mm' ).format( formatMoment );
            break;
        case "6am":
            eleMoment = moment( '06:00', 'hh:mm' ).format( formatMoment );
            break;

    }

    // Set Flatpickr
    switch ( pValue ) {
        case "Clear":
            document.querySelector( pIDFlatpickr )._flatpickr.clear( true );
            break;
        default:
            document.querySelector( pIDFlatpickr )._flatpickr.setDate( eleMoment, true, formatFlatpickr );
    }
}

function xanFileUploadToBucket( formSelector, postURL, callbackSuccess, callbackProblem ) {
    // Submit Directly
    // $( formSelector ).submit();
    // return;
    // Submit Directly
    let thisData = new FormData( $( formSelector )[ 0 ] );
    // alert( JSON.stringify( thisData ) );
    $.ajax( {
        type: "POST",
        url: postURL,
        data: thisData,
        cache: false,
        contentType: false,
        processData: false,
        success: function ( result, status, xhr ) {
            // alert( result.toString() );
            let jsonObj = JSON.parse( result );
            if ( jsonObj[ "IsOK" ] === 1 ) {
                let theURL = jsonObj[ "FilePathBase" ] + jsonObj[ "FileKey" ] + "/" + jsonObj[ "FileName" ];
                callbackSuccess( theURL );
            } else {
                callbackProblem( jsonObj[ "Message" ] );
            }
        },
        error: function ( xhr, status, error ) {
            // alert( error );
            callbackProblem( error );
        }
    } );
}

///////////////////////////////////////////////////////////
// Password Rating
// From https://github.com/caitlindaitch/passwordStrengthMeter with modifications of the class names.
//
// Calling example:
// var passwordRatingInput = document.getElementById('Password');
// xanPasswordRating( passwordRatingInput );
//
function xanPasswordRating( password ) {
    password.addEventListener( 'keyup', function () {

        // Init
        var passwordArray = password.value.split( '' );
        var totalScore = 0;
        var rating = {
            number: 0,
            lowercase: 0,
            uppercase: 0,
            specialChar: 0,
            lengthMin: 0,
            total: 0
        }

        // Validation Rules
        var validation = {
            isNumber: function ( val ) {
                var pattern = /^\d+$/;
                return pattern.test( val );
            },
            isLowercase: function ( val ) {
                var pattern = /[a-z]/;
                return pattern.test( val );
            },
            isUppercase: function ( val ) {
                var pattern = /[A-Z]/;
                return pattern.test( val );
            },
            isSpecialChar: function ( val ) {
                var pattern = /^[!@#\$%\^\&*\)\(+=._-]+$/g;
                return pattern.test( val );
            }
        }

        // Validate Chars
        for ( var i = 0; i < passwordArray.length; i++ ) {
            if ( validation.isNumber( passwordArray[ i ] ) ) {
                rating.number = 1;
            } else if ( validation.isLowercase( passwordArray[ i ] ) ) {
                rating.lowercase = 1;
            } else if ( validation.isUppercase( passwordArray[ i ] ) ) {
                rating.uppercase = 1;
            } else if ( validation.isSpecialChar( passwordArray[ i ] ) ) {
                rating.specialChar = 1;
            }
        }
        // Validate Length
        if ( passwordArray.length >= 10 ){
            rating.lengthMin = 1
        }
        if ( passwordArray.length >= 20 ){
            rating.lengthMin = 2
        }

        // Function Do
        function assessTotalScore() {
            var ratingElement = document.querySelector( ".passwordRating" );
            rating.total = rating.number + rating.lowercase + rating.uppercase + rating.specialChar + rating.lengthMin;

            if ( rating.total < 1 ) {
                ratingElement.innerHTML = "";
                ratingElement.classList.remove( "passwordWeak", "passwordModerate", "passwordStrong" );
            } else if ( rating.total === 1 ) {
                ratingElement.innerHTML = "&nbsp;Weak&nbsp;";
                ratingElement.classList.remove( "passwordModerate", "passwordStrong" );
                ratingElement.classList.add( "passwordWeak" );
            } else if ( rating.total === 2 ) {
                ratingElement.innerHTML = "&nbsp;Weak&nbsp;";
                ratingElement.classList.remove( "passwordModerate", "passwordStrong" );
                ratingElement.classList.add( "passwordWeak" );
            } else if ( rating.total === 3 ) {
                ratingElement.innerHTML = "&nbsp;Moderate&nbsp;";
                ratingElement.classList.remove( "passwordWeak", "passwordStrong" );
                ratingElement.classList.add( "passwordModerate" );
            } else if ( rating.total === 4 ) {
                ratingElement.innerHTML = "&nbsp;Moderate&nbsp;";
                ratingElement.classList.remove( "passwordWeak", "passwordStrong" );
                ratingElement.classList.add( "passwordModerate" );
            } else if ( rating.total > 4 ) {
                ratingElement.innerHTML = "&nbsp;Strong&nbsp;";
                ratingElement.classList.remove( "passwordWeak", "passwordModerate" );
                ratingElement.classList.add( "passwordStrong" );
            }
        }

        // Function Call
        assessTotalScore();

    } );
}

///////////////////////////////////////////////////////////
// Geolocation

function xanLocationGet( pCallbackFunction ) {
    // Calling Example
    // locationGet( function ( coords ) { alert( coords['ErrorCode'] + '|' + coords['Latitude'] ); } );
    if ( navigator.geolocation ) {
        navigator.geolocation.getCurrentPosition(
            function ( position ) {
                let result = {};
                result[ 'ErrorCode' ] = "0";
                result[ 'ErrorDesc' ] = "OK";
                result[ 'Latitude' ] = position.coords.latitude;
                result[ 'Longitude' ] = position.coords.longitude;
                result[ 'Altitude' ] = position.coords.altitude;
                result[ 'Heading' ] = position.coords.heading;
                result[ 'Speed' ] = position.coords.speed;
                result[ 'Accuracy' ] = position.coords.accuracy;
                result[ 'AccuracyAltitude' ] = position.coords.altitudeAccuracy;
                pCallbackFunction( result );
            },
            function ( err ) {
                let result = {};
                result[ 'ErrorCode' ] = err.code;
                result[ 'ErrorDesc' ] += ''; // err.message;
                result[ 'ErrorDesc' ] += ( err.code === 1 ) ? "The acquisition of the geolocation information failed because the page didn't have the permission to do it." : "";
                result[ 'ErrorDesc' ] += ( err.code === 2 ) ? "The acquisition of the geolocation failed. Using Safari on Mac OS? Enable System Preferences > Security and Privacy > Privacy > Location Services > Safari." : "";
                result[ 'ErrorDesc' ] += ( err.code === 3 ) ? "The time allowed to acquire the geolocation, was reached before the information was obtained." : "";
                pCallbackFunction( result );
            },
            {
                enableHighAccuracy: true,
                timeout: 15000,
                maximumAge: 0
            }
        )
    } else {
        let result = {};
        result[ 'ErrorCode' ] = -1;
        result[ 'ErrorDesc' ] = "Browser does not support Geolocation";
        callbackFunction( result );
    }
}

///////////////////////////////////////////////////////////
// Stripe

function xanStripeButtonInit( pStripeKeyPublic, pSessionID, pButtonID ) {
    let stripe = Stripe( pStripeKeyPublic );
    let stripeButton = document.getElementById( pButtonID );
    stripeButton.addEventListener( "click", function () {
        stripe.redirectToCheckout( {
            sessionId: pSessionID
        } ).then( function ( result ) {
            alert( "Error: " + result.error.message );
        } );
    } );
}